src/overlays/actors/ovl_En_Grasshopper/z_en_grasshopper.c

```diff
@@ -64,92 +138,990 @@ const ActorInit En_Grasshopper_InitVars = {
+            this->flyingHomePos.y = this->actor.floorHeight + 90.0f;
+            EnGrasshopper_ChangeAnim(this, EN_GRASSHOPPER_ANIM_FLY);
+            if (this->decision != EN_GRASSHOPPER_DECISION_ROAM_IN_CIRCLES) {
+                if (Player_GetMask(play) == PLAYER_MASK_STONE) {
+                    EnGrasshopper_SetupFly(this);
+                } else {
+                    EnGrasshopper_SetupApproachPlayer(this, play);
+                }
+            } else {
+                this->timer = 0;
+                this->action = EN_GRASSHOPPER_ACTION_ROAM_IN_CIRCLES;
+                this->decision = EN_GRASSHOPPER_DECISION_ATTACK;
+                this->waitTimer = this->timer;
+                this->actionFunc = EnGrasshopper_RoamInCircles;
+            }
+        }
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A6B6C.s")
+void EnGrasshopper_SetupFly(EnGrasshopper* this) {
+    EnGrasshopper_ChangeAnim(this, EN_GRASSHOPPER_ANIM_FLY);
+    this->baseFlyHeight = randPlusMinusPoint5Scaled(50.0f) + this->flyingHomePos.y;
+    this->action = EN_GRASSHOPPER_ACTION_FLY;
+    this->actionFunc = EnGrasshopper_Fly;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A6E18.s")
+void EnGrasshopper_Fly(EnGrasshopper* this, PlayState* play) {
+    f32 diffX;
+    f32 diffZ;
+    f32 targetSpeed;
+    Vec3f collisionCheckPos;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A6E74.s")
+    Actor_PlaySfxAtPos(&this->actor, NA_SE_EN_BATTA_FLY - SFX_FLAG);
+    diffX = this->flyingHomePos.x - this->actor.world.pos.x;
+    diffZ = this->flyingHomePos.z - this->actor.world.pos.z;
+    if ((this->type != EN_GRASSHOPPER_TYPE_WOODFALL_TEMPLE_FINAL_ROOM) &&
+        (this->type != EN_GRASSHOPPER_TYPE_WOODFALL)) {
+        this->bobPhase += 0xAF0;
+        this->targetPosY = (Math_SinS(this->bobPhase) * 10.0f) + this->baseFlyHeight;
+        Math_ApproachF(&this->actor.world.pos.y, this->targetPosY, 0.1f, 10.0f);
+        collisionCheckPos.x = (Math_SinS(this->actor.shape.rot.y) * 100.0f) + this->actor.world.pos.x;
+        collisionCheckPos.y = this->actor.world.pos.y;
+        collisionCheckPos.z = (Math_CosS(this->actor.shape.rot.y) * 100.0f) + this->actor.world.pos.z;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A6F8C.s")
+        if (this->collider.elements[0].info.ocElemFlags & OCELEM_HIT) {
+            this->shouldTurn = true;
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A700C.s")
+        if (BgCheck_SphVsFirstPoly(&play->colCtx, &collisionCheckPos, 10.0f)) {
+            this->shouldTurn = true;
+        } else {
+            this->shouldTurn = false;
+        }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Grasshopper/func_809A7134.s")
+        //! @bug: unreachable code. To get here, the type must NOT be EN_GRASSHOPPER_TYPE_WOODFALL
```

Comment:
```suggestion
        //! @bug Unreachable code. To get here, the type must NOT be EN_GRASSHOPPER_TYPE_WOODFALL
```
IIRC this is the usual format, although it's not very consistent.